gluten-flink/planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/stream/StreamExecExchange.java

```diff
@@ -0,0 +1,193 @@
+        KeySelector keySelector =
+            KeySelectorUtil.getRowDataSelector(
+                planner.getFlinkContext().getClassLoader(), keys, inputType);
+        // --- Begin Gluten-specific code changes ---
+        parallelism = inputTransform.getParallelism();
+        if (inputTransform instanceof OneInputTransformation) {
+          OneInputTransformation oneInputTransform = (OneInputTransformation) inputTransform;
+          if (oneInputTransform.getOperator() instanceof GlutenOperator) {
+            keySelector = new GlutenKeySelector();
+            final ExecEdge inputEdge = getInputEdges().get(0);
+            io.github.zhztheplayer.velox4j.type.RowType glutenInputType =
+                (io.github.zhztheplayer.velox4j.type.RowType)
+                    LogicalTypeConverter.toVLType(inputEdge.getOutputType());
+            io.github.zhztheplayer.velox4j.type.RowType outputType =
+                (io.github.zhztheplayer.velox4j.type.RowType)
+                    LogicalTypeConverter.toVLType(getOutputType());
+            String id = PlanNodeIdGenerator.newId();
+            List<Integer> keyIndexes = Arrays.stream(keys).boxed().collect(Collectors.toList());
+            PartitionFunctionSpec partitionFunctionSpec =
+                new HashPartitionFunctionSpec(glutenInputType, keyIndexes);
+            PlanNode localPartition =
+                new LocalPartitionNode(
+                    id,
+                    List.of(new EmptyNode(outputType)),
+                    "REPARTITION",
+                    false,
+                    partitionFunctionSpec);
+            PlanNode exchange = new StreamPartitionNode(id, localPartition, parallelism);
+            final GlutenSingleInputOperator exchangeKeyGenerator =
+                new GlutenSingleInputOperator(
+                    new StatefulPlanNode(id, exchange),
+                    id,
+                    glutenInputType,
+                    Map.of(id, outputType));
+            inputTransform =
+                ExecNodeUtil.createOneInputTransformation(
+                    inputTransform,
+                    new TransformationMetadata("exchange-hash", "Gluten exchange hash"),
+                    exchangeKeyGenerator,
+                    inputTransform.getOutputType(),
+                    parallelism,
+                    false);
+          }
+        }
+        partitioner =
+            new GlutenKeyGroupStreamPartitioner(keySelector, DEFAULT_LOWER_BOUND_MAX_PARALLELISM);
+        // --- End Gluten-specific code changes ---
+        break;
+      default:
+        throw new TableException(String.format("%s is not supported now!", distributionType));
+    }
+
+    final Transformation<RowData> transformation =
+        new PartitionTransformation<>(inputTransform, partitioner);
+    createTransformationMeta(EXCHANGE_TRANSFORMATION, config).fill(transformation);
+    transformation.setParallelism(parallelism);
```

Comment:
The parallelism set here is different from the original Flink code that uses default parallelism. Could you clarify it?